interpreter/runtimecost.go

```diff
@@ -0,0 +1,185 @@
+
+package interpreter
+
+import (
+	"math"
+
+	"github.com/google/cel-go/common/overloads"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+)
+
+// ActualCostEstimator provides function call cost estimations at runtime
+// CallCost returns an estimated cost for the function overload invocation with the given args, or nil if it has no
+// estimate to provide. CEL attempts to provide reasonable estimates for its standard function library, so CallCost
+// should typically not need to provide an estimate for CELs standard function.
+type ActualCostEstimator interface {
+	CallCost(overloadId string, args []ref.Val) *uint64
+}
+
+// CostObserver provides an observer that tracks runtime cost.
+func CostObserver(tracker *CostTracker) EvalObserver {
+	observer := func(id int64, programStep interface{}, val ref.Val) {
+		switch t := programStep.(type) {
+		case InterpretableConst:
+			// zero cost
+		case ConstantQualifier:
+			tracker.cost += 1
+		case InterpretableAttribute:
+			tracker.cost += 1
+		case Qualifier:
+			tracker.stack.pop()
+			tracker.cost += 1
+		case InterpretableCall:
+			argVals := make([]ref.Val, len(t.Args()))
+			argsFound := true
+			for i := len(t.Args()) - 1; i >= 0; i-- {
+				if v, ok := tracker.stack.pop(); ok {
+					argVals[i] = v
+				} else {
+					// should never happen
+					argsFound = false
+				}
+			}
+			if argsFound {
+				tracker.cost += tracker.costCall(t, argVals)
+			}
+		case InterpretableConstructor:
+			switch t.Type() {
+			case types.ListType:
+				tracker.cost += 10
+			case types.MapType:
+				tracker.cost += 30
+			default:
+				tracker.cost += 40
+			}
+		case InterpretableBooleanBinaryOp:
+			tracker.cost += 1
+		}
+		tracker.stack.push(val)
+		if tracker.Limit != nil && tracker.cost > *tracker.Limit {
+			panic(EvalCanceledError{Cause: ActualCostLimitExceeded})
+		}
+	}
+	return observer
+}
+
+// CostTracker represents the information needed for tacking runtime cost
+type CostTracker struct {
+	Estimator ActualCostEstimator
+	Limit     *uint64
+
+	cost  uint64
+	stack refValStack
+}
+
+// ActualCost returns the runtime cost
+func (c CostTracker) ActualCost() uint64 {
+	return c.cost
+}
+
+func (c CostTracker) costCall(call InterpretableCall, argValues []ref.Val) uint64 {
+	var cost uint64
+	if c.Estimator != nil {
+		callCost := c.Estimator.CallCost(call.OverloadID(), argValues)
+		if callCost != nil {
+			cost += *callCost
+			return cost
+		}
+	}
+	// if user didn't specify, the default way of calculating runtime cost would be used.
+	// if user has their own implementation of ActualCostEstimator, make sure to cover the mapping between overloadId and cost calculation
+	switch call.OverloadID() {
+	// O(n) functions
+	case overloads.StartsWithString, overloads.EndsWithString, overloads.StringToBytes, overloads.BytesToString:
+		cost += uint64(math.Ceil(float64(c.actualSize(argValues[0])) * 0.1))
```

Comment:
For these magic multipliers, I think it would be good to create a map of overload id to cost multiplier that can be shared between the `checker/cost.go` and the `interpreter/runtimecost.go`. At the very least creating a static map within each file (if they're different for some reason) seems helpful.

Reply:
I've broken them out as constants in a way that hopefully is reasonable.